File: src/etch/bindings/csharp/msg/Validator_int.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Etch.Msg
{
    /// <summary>
    /// Validator for int
    /// </summary>
    public class Validator_int : TypeValidator
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="nDims">number of dimensions. 0 for a scalar.</param>
        /// <returns>an instance of the validator</returns>
        public static Validator_int Get( int nDims )
        {
            CheckDims( nDims );

            if ( nDims >= validators.Length )
                return new Validator_int( nDims );

            Validator_int v = validators[ nDims ];

            if ( v == null )
                v = validators[ nDims ] = new Validator_int( nDims );

            return v;
        }

        private static Validator_int[] validators = new Validator_int[ MAX_CACHED ];

        /// <summary>
        /// Constructs the validator.
        /// </summary>
        /// <param name="nDims"></param>
        private Validator_int( int nDims )
            : base( typeof( int ), typeof( int ), nDims, "int["+nDims+"]" )
        { 
        }

        public override Validator ElementValidator()
        {
            return Get( nDims - 1 );
        }

        public override sbyte? CheckValue( object value )
        {
            if ( !Validate( value ) )
                return null;

            if ( nDims > 0 )
                return Etch.Transport.Fmt.TypeCode.ARRAY;

            int v = ( Int32 ) value;

            if ( v >= SByte.MinValue && v <= SByte.MaxValue )
                return Etch.Transport.Fmt.TypeCode.INT1;

            if ( v >= short.MinValue && v <= short.MaxValue )
                return Etch.Transport.Fmt.TypeCode.INT2;

            return Etch.Transport.Fmt.TypeCode.INT4;
        }
    }
}

File: src/etch/bindings/java/msg/Validator_int.java
package etch.bindings.java.msg;

import etch.bindings.java.transport.fmt.TypeCode;

/**
 * Validator for int.
 */
public class Validator_int extends TypeValidator
{
	/**
	 * @param nDims the number of dimensions. 0 for a scalar.
	 * @return an instance of the validator.
	 */
	public static Validator_int get( int nDims )
	{
		if (nDims < 0)
			return null;
		
		if (nDims >= validators.length)
			return new Validator_int( nDims );
		
		if (validators[nDims] == null)
			validators[nDims] = new Validator_int( nDims );
		
		return validators[nDims];
	}
	
	private static Validator_int[] validators = new Validator_int[4];
	
	/**
	 * Constructs the Validator.
	 *
	 * @param nDims the number of dimensions. 0 for a scalar.
	 */
	private Validator_int( int nDims )
	{
		super( Integer.class, int.class, nDims, "int["+nDims+"]" );
	}

	@Override
    public Validator elementValidator()
    {
	    return get( nDims-1 );
    }

	@Override
	public Byte checkValue( Object value )
	{
		if (!validate( value))
			return null;
		
		if (nDims > 0)
			return TypeCode.ARRAY;
		
		int v = (Integer) value;
		
		if (v >= Byte.MIN_VALUE && v <= Byte.MAX_VALUE)
			return TypeCode.INT1;
		
		if (v >= Short.MIN_VALUE && v <= Short.MAX_VALUE)
			return TypeCode.INT2;

		return TypeCode.INT4;
	}
}
